TropicalGeometry: new linear_space algorithm for triv val on QQ

YueRen opened the pull request:
First version for issue https://github.com/oscar-system/Oscar.jl/issues/3749

We probably need a similar function for polynomials

YueRen committed:
TropicalGeometry: new linear_space algorithm for triv val on QQ

TropicalGeometry: compute pluecker vector if asked and not cached

TropicalGeometry: added missing citation

TropicalGeometry: added non_redundant flag

TropicalGeometry: fixed bug in ensure_pluecker_data, removed print

TropicalGeoemtry: improved tests

TropicalGeometry: fixed bug pointed out by Max

TropicalGeometry: incorporated Max' suggestion for clearer code

TropicalGeometry: fixed tests

TropicalGeometry: further work on attributes of linear spaces

TropicalGeometry: fixed book tests

YueRen commented:
@lkastner @benlorenz I get errors in the tests because of `subdivision_of_points`.  Is there anything I have to fix in my code or can I just ignore that?

benlorenz commented:
> @lkastner @benlorenz I get errors in the tests because of `subdivision_of_points`. Is there anything I have to fix in my code or can I just ignore that?

The subdivision test should probably compare as sets instead of vectors, since the maximal cells are just reordered. Please try changing this test to:
```julia
@test issetequal(collect(maximal_cells(square_by_incidence)), collect(maximal_cells(square_by_weights)))
```
Or even
```julia
@test issetequal(maximal_cells(square_by_incidence), maximal_cells(square_by_weights))
```

Not sure why this happens though, how the new code might trigger these reorderings. It might also depend on the order in which all tests are run. There are some runs for 'short' where the test succeeded.
I will try looking into which test might touch some global state.

_Edit:_ the rebase and new merge with master might also make this disappear, then we can take care of this in a separate PR.

But there is another error in the booktests due to changed output where the output in the jlcon (probably `linearSpace1.jlcon`) needs to be adjusted:
```diff
  julia> R,(w,x,y,z) = QQ[\"w\",\"x\",\"y\",\"z\"];
  
  julia> nu = tropical_semiring_map(QQ);
  
  julia> I = ideal([w+x+y+z,w+2*x+5*y+11*z]);
  
  julia> TropL = tropical_linear_space(I,nu)
  Min tropical linear space
  
  julia> vertices_and_rays(TropL)
  5-element SubObjectIterator{Union{PointVector{QQFieldElem}, RayVector{QQFieldElem}}}:
+  [0, 1, 0, 0]
   [0, 0, 0, 0]
   [0, -1, -1, -1]
-  [0, 1, 0, 0]
   [0, 0, 1, 0]
   [0, 0, 0, 1]
  
  julia> maximal_polyhedra(IncidenceMatrix,TropL)
  4×5 IncidenceMatrix
  [1, 2]
- [1, 3]
- [1, 4]
- [1, 5]
+ [2, 3]
+ [2, 4]
+ [2, 5]
```

YueRen commented:
The pull request should be ready for review.  The code coverage goes down a bit, but currently it is impossible to call that portion of the code without removing attributes which I don't think is possible.

Is there a way for @oskarhenriksson to review the pull request?  The pull request fixes an issue of his.

thofma commented:
I think anyone with an account can submit a review in the "Files changed" window.

benlorenz commented:
The book test fails since it still needs to be adjusted for the re-ordered output.

YueRen commented:
@oskarhenriksson I have updated the tests and made sure the that tropical semiring map is an optional parameter (default is min-convention).  If you have the time (and once most tests have passed successfully), please review the changes!

oskarhenriksson commented:
@YueRen Thanks a lot! This seems to do what I want, and it seems like it's faster than the original `tropical_linear_space`.

oskarhenriksson commented:
A small concern though:

I just did some quick experiments, and it seems like a fairly major bottleneck in the new function is the computation of `n_maximal_polyhedra(Sigma)` in [line 293](https://github.com/oscar-system/Oscar.jl/blob/b0350459c34145173447ac8baf63458402295dcc/src/TropicalGeometry/linear_space.jl#L293C39-L293C65). 

For instance, if we take `A` to be a random 5-by-20 matrix with integer entries, the computation of `PolyhedralComplex{QQFieldElem}(Polymake.tropical.matroid_fan{convention(nu)}(A))` takes 4 seconds, whereas the computation of `n_maximal_polyhedra(Sigma)` takes 29 seconds. (For comparison, the whole computation with the old `tropical_linear_space` took 100 seconds.)

It feels like we already should have this information in the Polymake output from `matroid_fan`, so perhaps there could be a faster way to compute it?

oskarhenriksson commented on src/TropicalGeometry/linear_space.jl:
```diff
@@ -283,14 +284,27 @@ function tropical_linear_space(A::MatElem, nu::Union{Nothing,TropicalSemiringMap
+    Sigma = PolyhedralComplex{QQFieldElem}(Polymake.tropical.matroid_fan{convention(nu)}(A))
+    # Add the missing all ones vector to the lineality
+    Sigma = add_missing_lineality_from_polymake(Sigma)
+    multiplicities = ones(ZZRingElem, n_maximal_polyhedra(Sigma))
```
Perhaps one can consider the following instead (to avoid the costly computation of `n_maximal_polyhedra(Sigma)`). With this, the whole computation for a random 5-by-20 matrix `A` goes down from 30 to 4 seconds. (I'm not sure why this info isn't automatically carried over when we broadcast to `PolyhedralComplex{QQFieldElem}`?)

```suggestion
    # Compute Bergman fan in polymake and convert to polyhedral complex
    B = Polymake.tropical.matroid_fan{convention(nu)}(A)
    Sigma = PolyhedralComplex{QQFieldElem}(B)
    # Add the missing all ones vector to the lineality
    Sigma = add_missing_lineality_from_polymake(Sigma)
    multiplicities = ones(ZZRingElem, B.N_MAXIMAL_POLYTOPES)
```

benlorenz replied:
The extra time for `n_maximal_polyhedra` is probably redundancy elimination after adding the extra lineality. And by not doing it here it will just delay it to the first time the user triggers some other computation with the object.

If it is guaranteed that this lineality does not change the combinatorics (and is not implicit in the other lineality rows) then adding the internal `non_redundant` flag for the constructor in `add_missing_lineality_from_polymake` might be a better approach as polymake will then directly store the properties as irredundant description.

YueRen replied:
Done, thanks for the tip!  The `non_redundant` flat will come very useful for me.

oskarhenriksson commented on docs/src/TropicalGeometry/linear_space.md:
```diff
@@ -10,10 +10,12 @@ Objects of type `TropicalLinearSpace` need to be embedded, abstract tropical lin
+2. Pluecker vectors over a field and a tropical semiring map: computes coordinatewise valuation and uses constructor (1.)
+3. matrices over a tropical semiring: computes tropical minors and uses constructor (1.)
+4. matrices over a field and a tropical semiring map.
+    - if matrix over `QQ` and tropical semiring map is trivial, uses a fast specialized implementation in `polymake`
```
Perhaps it could make sense to [cite](https://github.com/oscar-system/Oscar.jl/blob/master/docs/oscar_references.bib) the [paper by Felipe Rincón](https://www.sciencedirect.com/science/article/pii/S0747717112001174?via%3Dihub) on the underlying algorithm here.

YueRen replied:
Good point, it's added in the version I just pushed now.

joschmitt commented:
> The code coverage goes down a bit, but currently it is impossible to call that portion of the code without removing attributes which I don't think is possible.

Stupid question: as I understand it, you say that it is impossible to create an example that calls those parts of the code. So why have this code? Do we expect it to be "callable" in the near future?

fingolfin commented on src/TropicalGeometry/linear_space.jl:
```diff
@@ -335,13 +349,72 @@ end
 function pluecker_indices(TropL::TropicalLinearSpace)
-    @req has_attribute(TropL,:pluecker_indices) "no pluecker indices cached"
+    ensure_pluecker_data!(TropL)
     return get_attribute(TropL,:pluecker_indices)
```
But `ensure_pluecker_data!` does not always set `pluecker_indices`, so shouldn't the `@req` stay, or a similar new test be added?


Looking at the code, how about something like this instead (ideally with a more specific return type for `algebraic_matrix`, but since you effectively use `Any` right now, I put that there):
```julia
@attr Vector{Vector{In64}} function pluecker_indices(TropL::TropicalLinearSpace)
    if has_attribute(TropL,:tropical_matrix)
        A = get_attribute(TropL,:tropical_matrix)
    else
        A = algebraic_matrix(TropL)
    end
    n = ncols(A)
    k = rank(A)
    return AbstractAlgebra.combinations(1:n,k)
end

@attr Any function algebraic_matrix(TropL::TropicalLinearSpace)
    I = get_attribute(TropL,:algebraic_ideal)
    @req I !== nothing "insufficient data to compute algebraic matrix"
    A = matrix([[coeff(g,xi) for xi in gens(base_ring(I))] for g in gens(I)])
    return A
end
```

I think you can rewrite the other function in a similar vein -- and then even dealing with the case I asked about elsewhere ("what if `tropical_matrix ` and `tropical_semiring_map` are known but none of `algebraic_pluecker_vector`, `algebraic_matrix`, `algebraic_ideal`?") becomes relatively simple.

YueRen replied:
Ah, you got me there.  That is a very good point, I will take a look at it later!

YueRen replied:
@fingolfin I like your suggestion, I'll make the necessary changes now!

YueRen replied:
One question: what is the `@attr` for in `@attr Any function algebraic_matrix(TropL::TropicalLinearSpace)`?

joschmitt replied:
`@attr` automatically enables attribute storage for the function. So the result is stored and looked up in the attribute `:algebraic_matrix`. See also the doc string for `@attr`.

fingolfin commented on src/TropicalGeometry/linear_space.jl:
```diff
@@ -335,13 +349,72 @@ end
+            k = rank(A)
+            plueckerIndices = AbstractAlgebra.combinations(1:n,k)
+            tropicalPlueckerVector = AbstractAlgebra.minors(A, k)
+            println("tropical matrix -> tropical pluecker")
```
```suggestion
```

fingolfin commented on src/TropicalGeometry/linear_space.jl:
```diff
@@ -335,13 +349,72 @@ end
+        end
+    else
+        # if not, we have to compute the tropical pluecker vector from the tropical data
+        if has_attribute(TropL,:tropical_matrix)
```
I find this function a bit difficult to untangle and I am not sure it is is always used right (see my other comments). In any case, isn't it true that if `tropical_semiring_map` is present, but none of `algebraic_pluecker_vector`, `algebraic_matrix`, `algebraic_ideal` are, then this method will fail -- even if `tropical_matrix` *is* available and thus would allow for the requested computation?

YueRen replied:
Good catch, that is very true and definitely shouldn't be the case!

YueRen replied:
Removing the `else` should do the trick, because if the data has been computed, we would have exited the function already.

YueRen commented:
> > The code coverage goes down a bit, but currently it is impossible to call that portion of the code without removing attributes which I don't think is possible.
> 
> Stupid question: as I understand it, you say that it is impossible to create an example that calls those parts of the code. So why have this code? Do we expect it to be "callable" in the near future?

It's so that I do not have to remember to implement that portion when it becomes callable (e.g., when removal of attributes is implemented, or a new constructor of tropical linear space makes that combination of attributes possible).

joschmitt commented:
> > > The code coverage goes down a bit, but currently it is impossible to call that portion of the code without removing attributes which I don't think is possible.
> > 
> > 
> > Stupid question: as I understand it, you say that it is impossible to create an example that calls those parts of the code. So why have this code? Do we expect it to be "callable" in the near future?
> 
> It's so that I do not have to remember to implement that portion when it becomes callable (e.g., when removal of attributes is implemented, or a new constructor of tropical linear space makes that combination of attributes possible).

Well, it is technically possible to remove attributes:
```
julia> R,(w,x,y,z) = QQ["w","x","y","z"];

julia> I = ideal([w+x+y+z]);

julia> TropL = tropical_linear_space(I);

julia> has_attribute(TropL, :algebraic_pluecker_vector)
true

julia> delete!(TropL.__attrs, :algebraic_pluecker_vector);

julia> has_attribute(TropL, :algebraic_pluecker_vector)
false
```
I wouldn't endorse this kind of stuff, but I think it's better than having untested code.

YueRen commented:
> > > > The code coverage goes down a bit, but currently it is impossible to call that portion of the code without removing attributes which I don't think is possible.
> > > 
> > > 
> > > Stupid question: as I understand it, you say that it is impossible to create an example that calls those parts of the code. So why have this code? Do we expect it to be "callable" in the near future?
> > 
> > 
> > It's so that I do not have to remember to implement that portion when it becomes callable (e.g., when removal of attributes is implemented, or a new constructor of tropical linear space makes that combination of attributes possible).
> 
> Well, it is technically possible to remove attributes:
> 
> ```
> julia> R,(w,x,y,z) = QQ["w","x","y","z"];
> 
> julia> I = ideal([w+x+y+z]);
> 
> julia> TropL = tropical_linear_space(I);
> 
> julia> has_attribute(TropL, :algebraic_pluecker_vector)
> true
> 
> julia> delete!(TropL.__attrs, :algebraic_pluecker_vector);
> 
> julia> has_attribute(TropL, :algebraic_pluecker_vector)
> false
> ```
> 
> I wouldn't endorse this kind of stuff, but I think it's better than having untested code.

Good point, I'll add it to the test later then.

YueRen commented:
I've squashed all commits and rebased to latest master.  Unless there are any objections, I think the pull request is ready to go once all tests are through.

lkastner commented on src/TropicalGeometry/linear_space.jl:
```diff
@@ -208,10 +209,25 @@ function tropical_linear_space(k::Int, n::Int, plueckerVector::Vector, nu::Union
     tropical_linear_space(A::MatElem{<:TropicalSemiringElem}; weighted_polyhedral_complex_only::Bool=false)
 
-Return a tropical linear space whose Pluecker vector are the tropical minors of `A`.  If `weighted_polyhedral_complex==true`, will not cache any extra information.
+Return a tropical linear space whose Pluecker vector are the tropical minors of `A`.  Assumes that `ncols(A)>=nrows(A)`.  If `weighted_polyhedral_complex==true`, will not cache any extra information.
```
You could `@req ncols(A)>=nrows(A)`, it is probably not very expensive.

YueRen replied:
Will be done!

lkastner commented on src/TropicalGeometry/linear_space.jl:
```diff
@@ -260,20 +273,17 @@ Min tropical linear space
-    isnothing(nu) && (nu=tropical_semiring_map(base_ring(A)))
+function tropical_linear_space(A::Union{MatElem,Matrix}, nu::Union{Nothing,TropicalSemiringMap}=nothing; weighted_polyhedral_complex_only::Bool=false)
+    # convert to Oscar matrix (necessary for AbstractAlgebra.minors)
+    A isa Matrix && (A = matrix(parent(first(A)),A))
```
Optionally:
```suggestion
tropical_linear_space(A::AbstractMatrix, nu::Union{Nothing,TropicalSemiringMap}=nothing; weighted_polyhedral_complex_only::Bool=false) =
    tropical_linear_space(matrix(parent(first(A)),A), nu; weighted_polyhedral_complex_only)
function tropical_linear_space(A::MatElem, nu::Union{Nothing,TropicalSemiringMap}=nothing; weighted_polyhedral_complex_only::Bool=false)
```
Just make sure to attach the docs to the right function, so probably the top function needs to be moved below.

YueRen replied:
I like your version better, I'll change it!

YueRen commented:
Everything should be good to go.  I also double-checked that the docu for the linear spaces looks correct, after incorporating the changes suggested by @lkastner.
